src/mono/wasm/runtime/binding_support.js

```diff
@@ -1752,6 +1779,238 @@ var BindingSupportLib = {
+			if (count > 0) {
+				const targetView = Module.HEAPU8.subarray(buffer_root.value, buffer_root.value + buffer_length);
+				const sourceView = event.data.subarray(event.offset, event.offset + count);
+				targetView.set(sourceView, 0);
+				event.offset += count;
+			}
+			const end_of_message = event.data.length === event.offset;
+			if (end_of_message) {
+				event_queue.dequeue();
+			}
+			return {
+				type: event.type,
+				count: count,
+				end_of_message: end_of_message
+			};
+		},
+
+		_text_decoder_utf8: undefined,
+		_mono_wasm_web_socket_send_buffering: function (ws, message_ptr, length, message_type, end_of_message) {
+			var buffer = ws[this.wasm_ws_pending_send_buffer];
+			var offset = 0;
+
+			if (buffer) {
+				offset = ws[this.wasm_ws_pending_send_buffer_offset];
+				// match desktop WebSocket behavior by copying message_type of the first part
+				message_type = ws[this.wasm_ws_pending_send_buffer_type];
+				// if not empty message, append to existing buffer
+				if (length !== 0) {
+					const view = Module.HEAPU8.subarray(message_ptr, message_ptr + length);
+					if (offset + length > buffer.length) {
+						const newbuffer = new Uint8Array((offset + length + 50) * 1.5); // exponential growth
+						newbuffer.set(buffer, 0, offset);// copy previous buffer
+						newbuffer.set(view, offset);// append copy at the end
+						ws[this.wasm_ws_pending_send_buffer] = buffer = newbuffer;
+					}
+					else {
+						buffer.set(view, offset);// append copy at the end
+					}
+					offset += length;
+					ws[this.wasm_ws_pending_send_buffer_offset] = offset;
+				}
+			}
+			else if (!end_of_message) {
+				// create new buffer
+				if (length !== 0) {
+					const view = Module.HEAPU8.subarray(message_ptr, message_ptr + length + 1024);
+					buffer = new Uint8Array(view); // copy
+					offset = length;
+					ws[this.wasm_ws_pending_send_buffer_offset] = offset;
+					ws[this.wasm_ws_pending_send_buffer] = buffer;
+				}
+				ws[this.wasm_ws_pending_send_buffer_type] = message_type;
+			}
+			else {
+				// use the buffer only localy
+				if (length !== 0) {
+					const memoryView = Module.HEAPU8.subarray(message_ptr, message_ptr + length);
+					buffer = memoryView; // send will make a copy
+					offset = length;
+				}
+			}
+			// buffer was updated, do we need to trim and convert it to final format ?
+			if (end_of_message) {
+				if (offset == 0) {
+					return new Uint8Array();
+				}
+				if (message_type === 0) {
+					// text, convert from UTF-8 bytes to string, because of bad browser API
+					if (this._text_decoder_utf8 === undefined) {
+						this._text_decoder_utf8 = new TextDecoder('utf-8');
+					}
+
+					// See https://github.com/whatwg/encoding/issues/172
+					var bytes = typeof SharedArrayBuffer !== 'undefined' && buffer instanceof SharedArrayBuffer
+						? buffer.slice(0, offset)
+						: buffer.subarray(0, offset);
+
+					return this._text_decoder_utf8.decode(bytes);
```

Comment:
Note that unlike the DOMString/USVString -> UTF8 encoder, the spec says that decode can throw an error. We should at minimum document this with a comment, but it would be nice to ensure that things behave relatively soundly if it throws (I don't think the current code would handle it at all.)

Reply:
You could also just state here that you're certain the decode won't fail given that we're encoding it on the .NET side (we just need to verify that the .NET encoder won't produce anything the JS decoder will reject, i.e. surrogate nonsense)

Reply:
I could not control if the C# user filled the buffers with correct UTF8, and so I rather enabled validation with `fatal: true`. The calling method would handle it and marshal back to C# side. I also wrote test for it.